keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1392 @@
+    <strong>Type of validation</strong>
+   </td>
+   <td>
+    <strong>IDL tag</strong>
+   </td>
+   <td>
+    <strong>Relative OpenAPI validation field</strong>
+   </td>
+  </tr>
+  <tr>
+   <td>
+    regex matches
+   </td>
+   <td>
+    <code>+k8s:pattern</code>
+   </td>
+   <td>
+    <code>+k8s:pattern</code>
+   </td>
+  </tr>
+  <tr>
+   <td>
+    cross field validation
+   </td>
+   <td>
+    <code>+k8s:validationRule="{CEL-like expression}"</code>
+   </td>
+   <td>
+    <code>+k8s:validationRule="{CEL-like expression}"</code>
+   </td>
+  </tr>
+  <tr>
+   <td>
+    <a href="https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules">transition rules</a>
+   </td>
+   <td>
+    <code>+k8s:validationRule="{CEL-like expression using oldSelf}"</code>
+   </td>
+   <td>
+    <code>+k8s:validationRule="{CEL-like expression using oldSelf}"</code>
+   </td>
+  </tr>
+  <tr>
+   <td>
+    special case: name format
+   </td>
+   <td>
+    <code>+k8s:nameFormat={format name}</code>
+   </td>
+   <td>
+    <code>+k8s:nameFormat={format name}</code>
+   </td>
+  </tr>
+</table>
+
+
+### Supporting Declarative Validation IDL tags On Shared Struct Fields
+
+
+IDL tags may be used directly on type declarations and indirectly on field and type aliases. For example:
+
+
+```go
+type Widget struct {
+  // +minLength=1
+  Component PartId json...
+}
+
+// +maxLength=20
+type PartId = Identifier
+
+// +format=dns1123label
+type Identifier string
+
+type Contraption struct {
+  Component Identifier json...
+}
+```
+
+
+
+In the above example, the `widget.component` field is validated against all three IDL tags. But `contraption.component` is only validated against `+format=dns1123label`.
+
+
+Shared types present a challenge. For example, different Kubernetes resources have different validation rules for `metadata.name` and `metadata.generateName`. But all resources share the `ObjectMeta` type.
+
+
+#### `subfield` IDL Tag
+
+
+To handle this case, we provide an IDL tag - `k8s:subfield(&lt;field-json-name>)` which can be used to specify a `subfield` validation to add to parent which validates against the the `subfield` value:
+
+
+```go
+type ExampleSpec struct {
+
+  // +k8s:subfield(name)=+k8s:+validationRule="!has(self.name) || self.name.isFormat('dns1123subdomain')",field="name"
+  metav1.ObjectMeta json...
+}
+```
+
+
+
+Because the above validation case of formatting `metav1.ObjectMeta.Name`  is so common, we are also considering creating additional tags to make declaring the above rules more convenient:
+
+
+```go
+// +nameFormat='dns1123subdomain'
+metav1.ObjectMeta json…
+```
+
+
+
+Another example: Pod `container` and `initContainer` fields share the same `Container` type but have different validation rules. The rules that are the same for both can be declared on `Container` but any rules that are different can be declared on the `container` and `initContainer` fields using CEL. E.g.:
```

Comment:
is the CEL stuff a carryover from previous rev of KEP?  I feel like it has not come up in other discussions?

Reply:
Yes apologies.  I’ve removed the CEL usage in all places from the document.  This was the remnants of a previous version of the document